### ISSUE TITLE
Run chef-solo via Ruby API

### ISSUE THREAD

**Issue by kowal** [chef/chef#3500]
Hi

I want to run chef-solo (and possibly other commands) via Ruby API. From what I see there's currently no way of setting options other than providing actual command line arguments (or changing ARGV). I want to run chef-solo from other command line app, so using ARGV it not an option.

What I did now is:
```ruby
class Chef::Application::Solo
  def parse_options(argv="--log_level debug".split)
    super(argv)
  end
end

Chef::Application::Solo.new.run
```

My question is - would it be possible to extract this as public API in ```Chef::Application``` to do something like this:
```ruby
Chef::Application::Solo.new(log_level: "debug").run
```

**Comment by ranjib** [comment]
I'll recommend  not to use the application class, as its intend to be consumed by the CLI directly. Instead, create /instantiate a chef client object, set apropriate configs (including Chef::Config[:solo] = true) and then invoke a chef run. Somewhat similar usage can be found here: https://github.com/ranjib/chef-lxc/blob/master/lib/chef/lxc_helper.rb#L8

**Comment by lamont-granquist** [comment]
You should probably just fork() + exec().  The cost of doing that is going to be astronomically small compared to the overall cost of the chef run itself.

**Comment by kowal** [comment]
Yup, we've already gave up on that idea. There was to many hacks involved. This was just initial idea that if we're already running Ruby app we should be able to call Chef via Ruby API. With that we could simplify error handling and make it easier to dynamically set options to chef-client run.

@ranjib thanks for the example - when I have more time for this, I'll try to call ``Chef::Client`` directly.

**Comment by lamont-granquist** [comment]
I'd strongly advise not building any new infrastructure around chef-solo as well.  You should build against chef-zero.  Eventually all the conditional code in the client around chef-solo is going to cause you trouble.

**Comment by kowal** [comment]
Good point about chef-zero.

Do you think I should switch to chef-zero even if I'm not interested in "chef server" architecture/workflow at this moment. Our infrastructure works fine with just chef-solo and I'm wondering what impact can chef-zero bring? Is it really make sense as target solution and not transition step?

**Comment by lamont-granquist** [comment]
Chef was designed around a server and chef-solo was a hack and always will be a hack and when you wind up going down code paths which are solo specific and hitting bugs you will be literally hosed.  We do not prioritize those.  The community support is also often not good.  Some bugs are architectural and cannot be fixed.

Knowing what we know now the way that chef-solo should have been written (to achieve the goal of a process that runs on a box without talking to an externally hosted server) would have been chef-zero.  We are using chef-zero extensively internally.  We have converted chef-solo cookbooks in the server over to using chef-zero.  We use chef-zero in our test-kitchen runs.  We use chef-zero internally in the integration tests that run in the chef-client and test it many times a day in travis.   Bugs in chef-zero will get prioritization and get fixed and there will likely be fewer of them because there are fewer codepaths in critical core sections of the client that are chef-zero-specific.

**Comment by kowal** [comment]
@lamont-granquist thanks for explanation. 

I'll close this as obviously this is not the way to go.

For anyone interested: in chef-solo workflow we're simply calling ```chef-solo``` binary via Ruby's ```Open3.popen3```. Not ideal but works.

At some point we're going to switch to ```chef-zero``` internally as we have our own command line interface on top of Chef.

thx

**Comment by mmzyk** [comment]
@kowal You've constructed your own command line interface on top of Chef? Would you be willing to chat with some of us from Chef about that at some point? We'd love to know what motivated you to do so and what your experience can tells us about Chef so we can improve our interfaces.

You can reach out to me at mm at chef.io if you you're willing to chat more. Thanks.

**Comment by kowal** [comment]
@mmzyk we are building our own CLI above Chef because it is being used by test engineers. The goal is to have git-like interface for setting up and testing various (local) environments:
```bash
# let's call this app 'run'
$ run prepare
$ run start
$ run test module_a
$ run test module_b
# etc
``` 
In some of those sub-commands we simply need to run chef-solo as we're implementing part of this as recipes.

One problem we came along was inability to pass JSON attributes on command line directly. 
```bash
# does not work
chef-solo --config solo.rb --log_level info --json-attributes '{run_list:..}'
```
We had to generate json temporary file before each run
```bash
chef-solo --config solo.rb --log_level info --json-attributes './tmp/another_one.json'
```

Basically this is the same issue as in https://tickets.opscode.com/browse/CHEF-1918. I agree that passing JSON over command line is not nice and what we did - generate temporary JSON before each run wasn't that hard.

I would appreciate any suggestions on how can we improve on that though :) Oh and maybe we can chat on https://gitter.im/chef/chef?

**Comment by kowal** [comment]
@morgoth , Wojtek I just run on https://tickets.opscode.com/browse/CHEF-1918 :) 5 years ago but maybe you remember if you came up with some better idea?

**Comment by morgoth** [comment]
@kowal Nope, unfortunately I don't remember. I wasn't using chef for a long time now.

**Comment by mmzyk** [comment]
@kowal I don't have any suggestions off hand, but I appreciate the feedback on what you're doing and why. It helps to inform what we might do to make things better going forward.

I'm not usually on glitter, but I can often be found on IRC as mmzyk.
